### PR TITLE
Pterodactyl 1.0

### LINKED ISSUE #41: Automatic firewall for CentOS (firewall-cmd)
**Is your feature request related to a problem? Please describe.**
A firewall is automatically configured for Ubuntu/Debian panel. It should be for CentOS as well.

**Describe the solution you'd like**
Implement the option to configure `firewall-cmd` as per the instructions in the README.

**Describe alternatives you've considered**
None.

**Additional context**
Not applicable.

### LINKED ISSUE #58: Automatic Let's Encrypt configuration for daemon
**Is your feature request related to a problem? Please describe.**
The installation script of the daemon does not create an SSL certificate.
There is still a query missing, which domain should be used incl. installing the certbot and issuing a certificate.
If you install everything on one system, it works without problems. If you separate panel and daemon, it will not work.

**Describe alternatives you've considered**
If i install certbot and create my own ssl-certificate all works fine :)

### LINKED ISSUE #61: Script stops when certbot cannon issue a certificate.
**Describe the bug**
When installing the panel, almost at the end of the script, if for some reason certbot fails to create the ssl the script just exit.

**To Reproduce**
Steps to reproduce the behaviour:
1. Run the script
2. Make certbot fails the cert
3. See error.

**Expected behaviour**
If certbot fails, the script should ask if the user still wants to continue the installation with ssl enabled or if he want it to fall back to http.

(I would love the option to continue assuming ssl, because for me certbot will always fail - not a bug, that's my server configuration)

**OS (please complete the following information):**
 - OS: Ubuntu 20.04
 - Installation: panel

### LINKED ISSUE #63: wings: Add link to configuration guide (official documentation)
At the end of the installation, you have to look around online for the steps in order to configure the daemon. I think it would be much easier if, at the end of the script, we added these lines (perhaps tweaking a little for coherence):

```
The guide to configuring your daemon can be found here: https://pterodactyl.io/daemon/0.6/installing.html#configure-daemon
Please run 'systemctl start wings' after the configuration.
```

### LINKED ISSUE #64: panel: Insert theme selector
TheFonix has some pretty good themes for the Pterodactyl panel available on [his GitHub repository](https://github.com/TheFonix/Pterodactyl-Themes). Below is a function that prompts the user which theme they want to install (if any), and performs the installation. It would be convenient for this function to be called somewhere in the installation script, eliminating the need for the themes to be installed separately from the panel.

```
pick_themes() {
	echo "Here, you can choose to install one of TheFonix's Pterodactyl themes."
	echo "Access previews and read more here: https://github.com/TheFonix/Pterodactyl-Themes"
	echo "These themes are not affiliated in any way with this installation script. Use at your own risk."
	echo "[0] Do not install any themes"
	echo "[1] Pink and Fluffy"
	echo "[2] Tango Twist"
	echo "[3] Blue Brick"
	echo "[4] Minecraft Madness"
	echo "[5] Lime Stitch"
	echo "[6] Red Ape"
	echo "[7] Black and the End of Space"
	echo "[8] Nothing but Graphite"
	cd /var/www/pterodactyl/
	while :
	do
		printf "Choose your theme: "
		read option
		case $option in
			0 ) echo "Keeping the vanilla Pterodactyl theme"
				break
				;;
			1 ) echo "Installing Pink and Fluffy"
				curl https://raw.githubusercontent.com/TheFonix/Pterodactyl-Themes/master/MasterThemes/PinkAnFluffy/build.sh | sh
				echo "Installed theme."
				break
				;;
			2 ) echo "Installing Tango Twist"
				invalid=false
				curl https://raw.githubusercontent.com/TheFonix/Pterodactyl-Themes/master/MasterThemes/TangoTwist/build.sh | sh
				echo "Installed theme."
				break
				;;
			3 ) echo "Installing Blue Brick"
				invalid=false
				curl https://raw.githubusercontent.com/TheFonix/Pterodactyl-Themes/master/MasterThemes/BlueBrick/build.sh | sh
				echo "Installed theme."
				break
				;;
			4 ) echo "Installing Minecraft Madness"
				invalid=false
				curl https://raw.githubusercontent.com/TheFonix/Pterodactyl-Themes/master/MasterThemes/MinecraftMadness/build.sh | sh
				echo "Installed theme."
				break
				;;
			5 ) echo "Installing Lime Stitch"
				invalid=false
				curl https://raw.githubusercontent.com/TheFonix/Pterodactyl-Themes/master/MasterThemes/LimeStitch/build.sh | sh
				echo "Installed theme."
				break
				;;
			6 ) echo "Installing Red Ape"
				invalid=false
				curl https://raw.githubusercontent.com/TheFonix/Pterodactyl-Themes/master/MasterThemes/RedApe/build.sh | sh
				echo "Installed theme."
				break
				;;
			7 ) echo "Installing Black and the End of Space"
				invalid=false
				curl https://raw.githubusercontent.com/TheFonix/Pterodactyl-Themes/master/MasterThemes/BlackEndSpace/build.sh | sh
				echo "Installed theme."
				break
				;;
			8 ) echo "Installing Nothing But Graphite"
				invalid=false
				curl https://raw.githubusercontent.com/TheFonix/Pterodactyl-Themes/master/MasterThemes/NothingButGraphite/build.sh | sh
				echo "Installed theme."
				break
				;;
			* ) echo "Invalid selection"
				;;
		esac
	done
}
```

### LINKED ISSUE #75: Adding the .env template and automatically filling it in.
**Is your feature request related to a problem? Please describe.**
Lots of people don't know what to select in the `php artisan p:environment:setup`, `php artisan p:environment:database`, `php artisan p:environment:mail` commands.

**Describe the solution you'd like**
Creating a .env file template, and replacing the crucial variables in it.

**Describe alternatives you've considered**
Not applicable.

### LINKED ISSUE #86: Let's Encrypt for CentOS
**Is your feature request related to a problem? Please describe.**
Centos doesn't have certbot, thus it is not possible to create certificates. 

**Describe the solution you'd like**
Add letsencrypt support to centOS

**Describe alternatives you've considered**
You can create certificates yourself.

**Additional context**
#85 Fixes

### LINKED ISSUE #89: panel: debian_based_letsencrypt function not found (pterodactyl-1.0)
`/dev/fd/62: line 811: debian_based_letsencrypt: command not found`
i installed certbot

### PR DESCRIPTION
**Pull request by vilhelmprytz**
Scripts adjusted for Pterodactyl 1.0 which is currently in beta. Will merge when Pterodactyl 1.0 releases!

Before merge: Revert the download and config URLs!!

**Approximate to-do list**

- [x] Try the script for all supported panel installations
    - [x] Ubuntu 18
    - [x] Ubuntu 20
    - [x] Debian 9
    - [x] Debian 10
    - [x] CentOS 7
    - [x] CentOS 8
- [x] Try the script for all supported wings installations
    - [x] Ubuntu 18
    - [x] Ubuntu 20
    - [x] Debian 9
    - [x] Debian 10
    - [x] CentOS 7
    - [x] CentOS 8
- [x] Update Wings to accommodate for the new default install path
- [x] Support CentOS 8
- [x] Initial rewrite for supporting panel
- [x] Initial rewrite for supporting Wings
- [x] wings: support Ubuntu 20.04
- [x] panel: support Ubuntu 20.04

Fixes #41 
Fixes #58 
Fixes #61 
Fixes #63
Fixes #64
Fixes #75
Fixes #86
Fixes #89

### CODE REVIEW

**Comment by VilhelmPrytz** [comment]
Should remove the external MariaDB repo on installations where it is possible (`focal` has `mariadb-server` package in default repo).

**Comment by VilhelmPrytz** [comment]
ubuntu 20: The `docker.io` package seems to be required.

**Comment by danielnaaman** [comment]
> ubuntu 20: The `docker.io` package seems to be required.

I believe that Docker isn't listed as supported yet on Focal Fossa, correct me if wrong.

https://docs.docker.com/engine/install/ubuntu/

**Comment by vilhelmprytz** [comment]
> > ubuntu 20: The `docker.io` package seems to be required.
> 
> I believe that Docker isn't listed as supported yet on Focal Fossa, correct me if wrong.
> 
> https://docs.docker.com/engine/install/ubuntu/
> 
> EDIT - it might be working, but it's not supported officially yet

I think you are very correct; not officially supported but reportedly works. We can hope Docker has full support before Pterodactyl 1.0 is out or else I'll just temporarily disable it :smiley:

**Comment by vilhelmprytz** [comment]
Experimental readded support for CentOS 8.

**Comment by vilhelmprytz** [comment]
Wings on Ubuntu 20 workaround: edit `/etc/apt/sources.list` and replace `focal` with `bionic` so that the Docker repo looks like this.
```
deb [arch=amd64] https://download.docker.com/linux/ubuntu bionic stable
```

Hopefully, Docker will support Ubuntu 20 soon!

**Comment by vilhelmprytz** [comment]
Wings: it seems like the latest beta release **only** looks for the config file in `/var/lib/pterodactyl/config.yml` which generates this error.

```bash
panic: stat /var/lib/pterodactyl/config.yml: no such file or directory

goroutine 1 [running]:
github.com/pterodactyl/wings/cmd.rootCmdRun(0x16d7be0, 0x1716278, 0x0, 0x0)
	cmd/root.go:72 +0x27ce
github.com/spf13/cobra.(*Command).execute(0x16d7be0, 0xc000084210, 0x0, 0x0, 0x16d7be0, 0xc000084210)
	/Users/dane/go/pkg/mod/github.com/spf13/cobra@v0.0.7/command.go:842 +0x29d
github.com/spf13/cobra.(*Command).ExecuteC(0x16d7be0, 0x0, 0xd2a3e0, 0xc000082058)
	/Users/dane/go/pkg/mod/github.com/spf13/cobra@v0.0.7/command.go:943 +0x317
github.com/spf13/cobra.(*Command).Execute(...)
	/Users/dane/go/pkg/mod/github.com/spf13/cobra@v0.0.7/command.go:883
github.com/pterodactyl/wings/cmd.Execute(...)
	cmd/root.go:220
main.main()
	wings.go:8 +0x2e
```

This is probably just a bug in the beta release; a temporary workaround is specifying the config file when starting Wings.

```bash
./wings --config /srv/wings/config.yml
```

**Comment by danielnaaman** [comment]
> Wings: it seems like the latest beta release **only** looks for the config file in `/var/lib/pterodactyl/config.yml` which generates this error.
> 
> ```shell
> panic: stat /var/lib/pterodactyl/config.yml: no such file or directory
> 
> goroutine 1 [running]:
> github.com/pterodactyl/wings/cmd.rootCmdRun(0x16d7be0, 0x1716278, 0x0, 0x0)
> 	cmd/root.go:72 +0x27ce
> github.com/spf13/cobra.(*Command).execute(0x16d7be0, 0xc000084210, 0x0, 0x0, 0x16d7be0, 0xc000084210)
> 	/Users/dane/go/pkg/mod/github.com/spf13/cobra@v0.0.7/command.go:842 +0x29d
> github.com/spf13/cobra.(*Command).ExecuteC(0x16d7be0, 0x0, 0xd2a3e0, 0xc000082058)
> 	/Users/dane/go/pkg/mod/github.com/spf13/cobra@v0.0.7/command.go:943 +0x317
> github.com/spf13/cobra.(*Command).Execute(...)
> 	/Users/dane/go/pkg/mod/github.com/spf13/cobra@v0.0.7/command.go:883
> github.com/pterodactyl/wings/cmd.Execute(...)
> 	cmd/root.go:220
> main.main()
> 	wings.go:8 +0x2e
> ```
> 
> This is probably just a bug in the beta release; a temporary workaround is specifying the config file when starting Wings.
> 
> ```shell
> ./wings --config /srv/wings/config.yml
> ```

See https://github.com/pterodactyl/wings/pull/30#issuecomment-626245688

**Comment by vilhelmprytz** [comment]
Ah, thank you very much. I have missed that they changed the default location for Wings. Will update the script accordingly.

**Comment by vilhelmprytz** [comment]
> > Wings: it seems like the latest beta release **only** looks for the config file in `/var/lib/pterodactyl/config.yml` which generates this error.
> > ```shell
> > panic: stat /var/lib/pterodactyl/config.yml: no such file or directory
> > 
> > goroutine 1 [running]:
> > github.com/pterodactyl/wings/cmd.rootCmdRun(0x16d7be0, 0x1716278, 0x0, 0x0)
> > 	cmd/root.go:72 +0x27ce
> > github.com/spf13/cobra.(*Command).execute(0x16d7be0, 0xc000084210, 0x0, 0x0, 0x16d7be0, 0xc000084210)
> > 	/Users/dane/go/pkg/mod/github.com/spf13/cobra@v0.0.7/command.go:842 +0x29d
> > github.com/spf13/cobra.(*Command).ExecuteC(0x16d7be0, 0x0, 0xd2a3e0, 0xc000082058)
> > 	/Users/dane/go/pkg/mod/github.com/spf13/cobra@v0.0.7/command.go:943 +0x317
> > github.com/spf13/cobra.(*Command).Execute(...)
> > 	/Users/dane/go/pkg/mod/github.com/spf13/cobra@v0.0.7/command.go:883
> > github.com/pterodactyl/wings/cmd.Execute(...)
> > 	cmd/root.go:220
> > main.main()
> > 	wings.go:8 +0x2e
> > ```
> > 
> > 
> > This is probably just a bug in the beta release; a temporary workaround is specifying the config file when starting Wings.
> > ```shell
> > ./wings --config /srv/wings/config.yml
> > ```
> 
> See [pterodactyl/wings#30 (comment)](https://github.com/pterodactyl/wings/pull/30#issuecomment-626245688)

Thanks @danielnaaman, now fixed.

File: install-wings.sh [hunk 1]
```diff
@@ -319,20 +302,6 @@ function systemd_file {
   echo "* Installed systemd service!"
 }
 
-function install_standalone_sftp_server {
```
Comment (vilhelmprytz) [inline install-wings.sh:322]: As far as I know, this is no longer required for the Wings.

**Comment by FoksVHox** [comment]
About the `*` in your todo list. I don't think you should keep support for those since the 1.0 isn't supporting them, which can cause those things that work on a 1.0 Panel on a 20.04 machine not to work on a 16.04 machine.

**Comment by danielnaaman** [comment]
> About the `*` in your todo list. I don't think you should keep support for those since the 1.0 isn't supporting them, which can cause those things that work on a 1.0 Panel on a 20.04 machine not to work on a 16.04 machine.

I tend to agree with this, the installer should just say that this version is unsupported by Pterodactyl and you should use it at your own risk, assuming it will even work properly.

On that note, I'd be happy to help with testing if the need exists.

**Comment by vilhelmprytz** [comment]
@FoksVHox @danielnaaman Thank you for your feedback regarding this! I am also thinking it might be best to not bother supporting it. If anyone wants to test the CentOS installations I'd be happy to review feedback here.

**Comment by FoksVHox** [comment]
@vilhelmprytz Also if you were to support the Debian 8 for the panel, it wouldn't work for the daemon. Since your todo list don't show it :)

**Comment by vilhelmprytz** [comment]
Support for older distros has now been removed.

@FoksVHox Regarding Debian 8: Debian 8 is not supported for daemon in `master` branch right now either (I think it was because of Docker).

**Comment by danielnaaman** [comment]
May I ask why the "--force" flag? It sounds risky.

Additionally, is it not possible to input "yes" by default, preventing this query in the first place?

**Comment by vilhelmprytz** [comment]
> May I ask why the "--force" flag? It sounds risky.
> 
> Additionally, is it not possible to input "yes" by default, preventing this query in the first place?

@danielnaaman Well, it is risky if you are running it on an already existing database since it will wipe anything that's already there. Because the script runs everything in a specific order we can _almost_ be sure that the database is empty when the database migration command is run.

And what do you mean with "not possible to input yes"?

**Comment by danielnaaman** [comment]
> > May I ask why the "--force" flag? It sounds risky.
> > Additionally, is it not possible to input "yes" by default, preventing this query in the first place?
> 
> @danielnaaman Well, it is risky if you are running it on an already existing database since it will wipe anything that's already there. Because the script runs everything in a specific order we can _almost_ be sure that the database is empty when the database migration command is run.
> 
> And what do you mean with "not possible to input yes"?

Gotcha.

Additionally,

There is a query to migrate the database, asking you to type yes, and considering yes is the only possible response, and no will cancel the installation, wouldn't it be smarter to just run it with a --yes flag if such is available?

Since this is an auto-installer, queries like this should be automated. (Ones that there's only one response that will assure installation)

File: install-panel.sh [hunk 2]
```diff
@@ -227,8 +224,7 @@ function configure {
   php artisan p:environment:mail
 
   # configures database
-  print_warning "You must type 'yes' or else the installer will fail! The default response 'no' will not properly initialize the database!"
-  php artisan migrate --seed
+  php artisan migrate --seed --force
```
Comment (vilhelmprytz) [inline install-panel.sh:227]: @danielnaaman Is it this step you are referring to? This is where I added the `--force` flag which will skip the `yes` prompt.

Comment (danielnaaman) [reply install-panel.sh:227]: > @danielnaaman Is it this step you are referring to? This is where I added the `--force` flag which will skip the `yes` prompt.

My bad haha

**Comment by vilhelmprytz** [comment]
panel: Ubuntu 18 and 20 now tested and works. Turns out MariaDB versions older than 10.2 were not supported so I fixed that for Ubuntu 18 (and for Ubuntu 9, yet to test though).

**Comment by vilhelmprytz** [comment]
panel/wings: Debian 10 tested to work

**Comment by vilhelmprytz** [comment]
Todo: update PR from master, conflicts from #62 must be solved

**Comment by vilhelmprytz** [comment]
The new script has now been tested to work with CentOS 7 as well. All supported installations are now tested at least once.

**Comment by FoksVHox** [comment]
This might be offtopic, but is there a reason that you choose Bash over something like Laravel Zero, for this?

**Comment by vilhelmprytz** [comment]
> This might be offtopic, but is there a reason that you choose Bash over something like Laravel Zero, for this?

I don't think you can run `apt` commands or download nginx configurations from Laravel Zero. I believe bash is the go-to standard here.

**Comment by FoksVHox** [comment]
> > This might be offtopic, but is there a reason that you choose Bash over something like Laravel Zero, for this?
> 
> I don't think you can run `apt` commands or download nginx configurations from Laravel Zero. I believe bash is the go-to standard here.

Ah okay. Tho it is kind of a bummer since I'd like to help with it, but don't know Bash

**Comment by vilhelmprytz** [comment]
> > > This might be offtopic, but is there a reason that you choose Bash over something like Laravel Zero, for this?
> > 
> > I don't think you can run `apt` commands or download nginx configurations from Laravel Zero. I believe bash is the go-to standard here.
> 
> Ah okay. Tho it is kind of a bummer since I'd like to help with it, but don't know Bash

@FoksVHox thankfully, bash is quite easy to learn. About Laravel Zero, it's for based on Laravel but for building command line applications, right? How is that even related to an installation script like this? Frankly, I think bash is the only "language" that would work for the script. 🙂

**Comment by danielnaaman** [comment]
> > > > This might be offtopic, but is there a reason that you choose Bash over something like Laravel Zero, for this?
> > > 
> > > 
> > > I don't think you can run `apt` commands or download nginx configurations from Laravel Zero. I believe bash is the go-to standard here.
> > 
> > 
> > Ah okay. Tho it is kind of a bummer since I'd like to help with it, but don't know Bash
> 
> @FoksVHox thankfully, bash is quite easy to learn. About Laravel Zero, it's for based on Laravel but for building command line applications, right? How is that even related to an installation script like this? Frankly, I think bash is the only "language" that would work for the script. slightly_smiling_face

Actually, I think Go and Python could do the trick and allow for more complicated installers and experiences, but honestly, I think this is the best implementation of the installer I've seen.

**Comment by adrijanb** [comment]
Hey,

somehow it is not selection 1.0

**Comment by vilhelmprytz** [comment]
> Hey,
> 
> somehow it is not selection 1.0

You'll have to be more specific, I don't understand your issue.

**Review by Linux123123 — APPROVED** [review]
Chnge the download urls and merge!!!!!!!

**Comment by vilhelmprytz** [comment]
> Chnge the download urls and merge!!!!!!!

Will do!

**Pull request merged**